src/extension-runners/firefox-android.js

```diff
@@ -0,0 +1,653 @@
+    );
+    await adbClient.startActivity(selectedAdbDevice.id, {
+      wait: true,
+      action: 'android.activity.MAIN',
+      component: `${selectedFirefoxApk}/.App`,
+      extras: [
+        {
+          key: 'args',
+          value: `-profile ${deviceProfileDir}`,
+        },
+      ],
+    });
+  }
+
+  async buildAndPushExtension(sourceDir: string) {
+    const {
+      adbClient,
+      selectedAdbDevice,
+      selectedArtifactsDir,
+      params: {
+        buildSourceDir,
+      },
+    } = this;
+
+    const {extensionPath} = await buildSourceDir(sourceDir);
+
+    const extFileName = path.basename(extensionPath, '.zip');
+
+    let adbExtensionPath = this.adbExtensionsPathBySourceDir.get(sourceDir);
+
+    if (!adbExtensionPath) {
+      adbExtensionPath = `${selectedArtifactsDir}/${extFileName}.xpi`;
+    }
+
+    log.debug(`Uploading ${extFileName} on the android device`);
+
+    await adbClient.push(
+      selectedAdbDevice.id, extensionPath, adbExtensionPath
+    ).then(function(transfer) {
+      return new Promise((resolve) => {
+        // TODO: show progress in the console
+        transfer.on('end', resolve);
+      });
+    });
+
+    log.debug(`Upload completed: ${adbExtensionPath}`);
+
+    this.adbExtensionsPathBySourceDir.set(sourceDir, adbExtensionPath);
+  }
+
+  async buildAndPushExtensions() {
+    for (const {sourceDir} of this.params.extensions) {
+      await this.buildAndPushExtension(sourceDir);
+    }
+  }
+
+  async adbDiscoveryAndForwardRDPUnixSocket() {
+    const {
+      adbClient,
+      selectedAdbDevice,
+      selectedFirefoxApk,
+      params: {
+        adb,
+        firefoxAndroidTimeout,
+      },
+    } = this;
+
+    const stdin = this.params.stdin || process.stdin;
+
+    const {
+      unixSocketDiscoveryRetryTime,
+    } = FirefoxAndroidExtensionRunner;
+
+    let {
+      unixSocketDiscoveryMaxTime,
+    } = FirefoxAndroidExtensionRunner;
+
+    if (typeof firefoxAndroidTimeout === 'number') {
+      unixSocketDiscoveryMaxTime = firefoxAndroidTimeout;
+    }
+
+    // Firefox for Android debugger socket discovery.
+    const startedDiscoveryTime = Date.now();
+    let androidUnixSockets = [];
+    let userExit = false;
+
+    const handleCtrlC = (str, key) => {
+      if (key.ctrl && key.name === 'c') {
+        userExit = true;
+      }
+    };
+
+    if (stdin.isTTY && stdin instanceof tty.ReadStream) {
+      readline.emitKeypressEvents(stdin);
+      stdin.setRawMode(true);
+
+      stdin.on('keypress', handleCtrlC);
+    }
+
+    while (!userExit && androidUnixSockets.length === 0) {
+      if (Date.now() - startedDiscoveryTime > unixSocketDiscoveryMaxTime) {
+        throw new WebExtError(
+          'Timeout while waiting for the Android Firefox Debugger Socket'
+        );
+      }
+
+      await new Promise((resolve) => {
+        setTimeout(resolve, unixSocketDiscoveryRetryTime);
+      });
+
+      log.info(`Waiting ${selectedFirefoxApk} to be ready... ` +
```

Comment:
I thought it was a bit annoying to see this message printed repeatedly so I made a little loading indicator if you want to use it. It might be hard to test though. Also, I think the modulo math is off by one maybe.

```diff
diff --git a/src/extension-runners/firefox-android.js b/src/extension-runners/firefox-android.js
index fae5b61..dfbd7d3 100644
--- a/src/extension-runners/firefox-android.js
+++ b/src/extension-runners/firefox-android.js
@@ -540,29 +540,50 @@ export class FirefoxAndroidExtensionRunner {
       stdin.on('keypress', handleCtrlC);
     }
 
-    while (!userExit && androidUnixSockets.length === 0) {
-      if (Date.now() - startedDiscoveryTime > unixSocketDiscoveryMaxTime) {
-        throw new WebExtError(
-          'Timeout while waiting for the Android Firefox Debugger Socket'
-        );
-      }
 
-      await new Promise((resolve) => {
-        setTimeout(resolve, unixSocketDiscoveryRetryTime);
-      });
+    const loadingPositions = ['---', ' / ', ' | ', ' \ '];
+    log.info(`Waiting ${selectedFirefoxApk} to be ready... `);
+    let loadingPos = 0;
+
+    const loader = setInterval(() => {
+      // Show a little loading indicator.
+      process.stdout.write(
+        `\r${loadingPositions[loadingPos % (loadingPositions.length - 1)]}`);
+      loadingPos += 1;
+    }, 200);
 
-      log.info(`Waiting ${selectedFirefoxApk} to be ready... ` +
-               '(it can take a while on an ARM emulator)');
-      androidUnixSockets = await adbClient.shell(selectedAdbDevice.id, [
-        'cat', '/proc/net/unix',
-      ]).then(adb.util.readAll);
+    function clearLoader() {
+      clearInterval(loader);
+    }
 
-      androidUnixSockets = androidUnixSockets.toString()
-        .split('\n').filter((line) => {
-          return line.trim().endsWith('firefox-debugger-socket');
+    while (!userExit && androidUnixSockets.length === 0) {
+      try {
+        if (Date.now() - startedDiscoveryTime > unixSocketDiscoveryMaxTime) {
+          throw new WebExtError(
+            'Timeout while waiting for the Android Firefox Debugger Socket'
+          );
+        }
+
+        await new Promise((resolve) => {
+          setTimeout(resolve, unixSocketDiscoveryRetryTime);
         });
+
+        androidUnixSockets = await adbClient.shell(selectedAdbDevice.id, [
+          'cat', '/proc/net/unix',
+        ]).then(adb.util.readAll);
+
+        androidUnixSockets = androidUnixSockets.toString()
+          .split('\n').filter((line) => {
+            return line.trim().endsWith('firefox-debugger-socket');
+          });
+      } catch (error) {
+        clearLoader();
+        throw error;
+      }
     }
 
+    clearLoader();
+
     if (stdin.isTTY && stdin instanceof tty.ReadStream) {
       stdin.removeListener('keypress', handleCtrlC);
     }
```